pkg/common/idutil/spiffeid.go

```diff
@@ -3,137 +3,24 @@ package idutil
-// - path does not start with '/spire' since it is reserved for agent, server, etc.
-// In addition, the validation mode is used to control what kind of SPIFFE ID
-// is expected.
-// For more information:
-// [https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md]
-func ValidateSpiffeIDURL(id *url.URL, mode ValidationMode) error {
-	options := mode.validationOptions()
-
-	validationError := func(format string, args ...interface{}) error {
-		var kind string
-		switch options.idType {
-		case trustDomainID:
-			kind = "trust domain "
-		case memberID:
-			kind = "trust domain member "
-		case workloadID:
-			kind = "workload "
-		case serverID:
-			kind = "server "
-		case agentID:
-			kind = "agent "
-		}
-		return fmt.Errorf("%q is not a valid %sSPIFFE ID: "+format,
-			append([]interface{}{id.String(), kind}, args...)...)
-	}
-
-	if id == nil || *id == (url.URL{}) {
-		return validationError("SPIFFE ID is empty")
-	}
-
-	// General validation
-	switch {
-	case strings.ToLower(id.Scheme) != "spiffe":
-		return validationError("invalid scheme")
-	case id.User != nil:
-		return validationError("user info is not allowed")
-	case id.Host == "":
-		return validationError("trust domain is empty")
-	case id.Port() != "":
-		return validationError("port is not allowed")
-	case id.Fragment != "":
-		return validationError("fragment is not allowed")
-	case id.RawQuery != "":
-		return validationError("query is not allowed")
-	}
-
-	// trust domain validation
-	if options.trustDomainRequired {
-		if options.trustDomain.IsZero() {
-			return errors.New("trust domain to validate against cannot be empty")
-		}
-		if id.Host != options.trustDomain.String() {
-			return fmt.Errorf("%q does not belong to trust domain %q", id, options.trustDomain)
-		}
+func MemberFromString(td spiffeid.TrustDomain, s string) (spiffeid.ID, error) {
+	id, err := spiffeid.FromString(s)
+	if err != nil {
+		return spiffeid.ID{}, err
 	}
-
-	// id type validation
-	switch options.idType {
-	case anyID:
-	case trustDomainID:
-		if id.Path != "" {
-			return validationError("path is not empty")
-		}
-	case memberID:
-		if id.Path == "" {
-			return validationError("path is empty")
-		}
-	case workloadID:
-		if id.Path == "" {
-			return validationError("path is empty")
-		}
-		if IsReservedPath(id.Path) {
-			return validationError(`invalid path: "/spire/*" namespace is reserved`)
-		}
-	case serverID:
-		if id.Path == "" {
-			return validationError("path is empty")
-		}
-		if !isServerPath(id.Path) {
-			return validationError(`invalid path: expecting "/spire/server"`)
-		}
-	case agentID:
-		if id.Path == "" {
-			return validationError("path is empty")
-		}
-		if !IsAgentPath(id.Path) {
-			return validationError(`invalid path: expecting "/spire/agent/*"`)
-		}
-	default:
-		return validationError("internal error: unhandled id type %v", options.idType)
+	if !id.MemberOf(td) {
+		return spiffeid.ID{}, fmt.Errorf("id %q is not a member of trust domain %q", id, td)
```

Comment:
Maybe we should explicit to say SPIFFE ID rather than just ID?

Reply:
i thought the linter would complain. it didn't. changed!